NEXT-12727 - Add guide on media and thumbnails (#126)
shopware/docs: guides/plugins/plugins/storefront/use-media-thumbnails.md

> # working with media and thumbnails

Reviewer: Uppercase w?

Author: Right 👍

> # working with media and thumbnails

Author: ```suggestion
Working with media and thumbnails
```

## Overview

In Shopware's Storefront, you can assign media objects to the different entities. To name an example, this is often 
used for products to show more information with images on the product detail page. This guide should give you a 
starting point on how to use media and thumbnails in your Storefront plugin.

## Prerequisites

In order to use your own media files or thumbnails of your plugin in the storefront, of course you first need a 
plugin as base. To create an own plugin, you can refer to the [Plugin Base Guide](./../plugin-base-guide.md).

Displaying custom images is often done by using custom fields. To take full advantage of this guide, you might want to
read the corresponding guide on [using custom fields](./../administration/add-custom-field.md).

## Using searchMedia function

> You're able to store media in your shop and maintaining them in your Administration. It is not possible to display 

Reviewer: > maintaining
should be
> maintain

> You're able to store media in your shop and maintaining them in your Administration. It is not possible to display 

Author: ```suggestion
You should be able to store media in your shop and to maintain them in your Administration. It is not possible to display 
```

such an image in the storefront with only its media ID though. To achieve that, the function `searchMedia` exists:

```php
public function searchMedia (array $ids, Context $context): MediaCollection { 
... 
}
```

This `searchMedia` function reads out the corresponding media objects for the given IDs in order to continue working with them 
afterwards. Here is an example with a custom field (`custom_sports_media_id`) on the product detail page:

```twig
{% sw_extends '@Storefront/storefront/page/product-detail/index.html.twig' %}

{% block page_product_detail_media %}
    {# simplify ID access #}
    {% set sportsMediaId = page.product.translated.customFields.custom_sports_media_id %}

    {# fetch media as batch - optimized for performance #}
    {% set mediaCollection = searchMedia([sportsMediaId], context.context) %}

    {# extract single media object #}
    {% set sportsMedia = mediaCollection.get(sportsMediaId) %}

    {{ dump (sportsMedia) }}
{% endblock %}
```

> {% hint style="danger" %} Please note that this function performs a query against the database and should therefore 
> not be used within a loop. {% endhint %}

Reviewer 2: ```suggestion
{% hint style="danger" %}
Please note that this function performs a query against the database and should therefore 
not be used within a loop.
{% endhint %}
```

To read the media objects within the product listing we recommend the following procedure:

```twig
{% sw_extends '@Storefront/storefront/component/product/listing.html.twig' %}

{% block element_product_listing_col %}
    {# initial ID array #}
    {% set sportsMediaIds = [] %}

    {% for product in searchResult %}
        {# simplify ID access #}
        {% set sportsMediaId = product.translated.customFields.custom_sports_media_id %}

        {# merge IDs to a single array #}
        {% set sportsMediaIds = sportsMediaIds|merge([sportsMediaId]) %}
    {% endfor %}

    {# do a single fetch from database #}
    {% set mediaCollection = searchMedia(sportsMediaIds, context.context) %}

    {% for product in searchResult %}
        {# simplify ID access #}
        {% set sportsMediaId = product.translated.customFields.custom_sports_media_id %}

        {# get access to media of product #}
        {% set sportsMedia = mediaCollection.get(sportsMediaId) %}

        {{ dump(sportsMedia) }}
    {% endfor %}
{% endblock %}
```

## Working with sw_thumbnail 

A common issue when developing responsive web pages is resizing images properly for different screen widths. By default,

> Shopware generates various thumbnails for each uploaded image. Normally you would have to manually write larger 

Reviewer 2: ```suggestion
Shopware generates various thumbnails for each uploaded image. Normally you would have to manually write large
```

chunks of HTML code to render the needed images with `img` and `srcset`.

Fortunately, you do not need to define these attributes on your own - For that, Shopware 
introduced the `sw_thumbnails` Twig function: `sw_thumbnails` automatically generates the `img` and `srcset` code.
This is the minimal configuration:
```twig
{% sw_thumbnails 'my-thumbnails' with {
    media: cover
} %}
```

As you see, `sw_thumbnail` makes use of one required parameter: `media` is required and contains the whole media entity.
The string after `sw_thumbnails` is also required but does not render a CSS class. All other parameters are optional.

### Dealing with thumbnail sizes

> With the `sizes` parameter you can control the `sizes` Attribute of the `img` and define which of the thumbnails should 

Reviewer 3: ```suggestion
With the `sizes` parameter you can control the `sizes` attribute of the `img` and define which of the thumbnails should 
```

           (max-width: 575px) and (min-width: 0px) 501px, 100vw">
```

By giving the `default` size you can override the media queries and always refer to a single image source for 
all viewports. To give an example, think about always using a small thumbnail closest to 100px regardless of the 
current viewport:

```twig
{% sw_thumbnails 'my-thumbnails' with {
    media: cover,
    sizes: {
        'xs': '501px', {# Will be ignored #}
        'sm': '315px', {# Will be ignored #}
        'md': '427px', {# Will be ignored #}
        'lg': '333px', {# Will be ignored #}
        'xl': '284px', {# Will be ignored #}
        'default': '100px'
    }
} %}
```

This example will create the output below:
```html
<img 
    src="http://shopware.local/media/06/f0/5c/1614258798/example-image.jpg" 
    srcset="http://shopware.local/media/06/f0/5c/1614258798/example-image.jpg 1921w, 
            http://shopware.local/thumbnail/06/f0/5c/1614258798/example-image_1920x1920.jpg 1920w, 
            http://shopware.local/thumbnail/06/f0/5c/1614258798/example-image_800x800.jpg 800w, 
            http://shopware.local/thumbnail/06/f0/5c/1614258798/example-image_400x400.jpg 400w" 
    sizes="100px">
```

> {% hint style="danger" %} Please note that those sizes only work with bootstrap viewports, like xs, sm, md, lg and xl.
> Custom media queries will not work. {% endhint %}

Reviewer 2: ```suggestion
{% hint style="danger" %}
Please note that those sizes only work with bootstrap viewports, like xs, sm, md, lg and xl.
Custom media queries will not work.
{% endhint %}
```

### Additional attributes

With the `attributes` param, additional attributes can be applied. Imagine the following example:
```twig
{% sw_thumbnails 'my-thumbnails' with {
    media: cover,
    attributes: {
        'class': 'my-custom-class',
        'alt': 'Doggo with cat',
        'title': 'Doggo with cat'
    }
} %}
```

This will generate the output below:
```html
<img 
    src="..." 
    sizes="..." 
    class="my-custom-class" 
    alt="Image name" 
    title="My beautiful image">

> ```

Reviewer 2: Maybe add a "## Next steps" section

# working with media and thumbnails

## Overview

In Shopware's Storefront, you can assign media objects to the different entities. To name an example, this is often 
used for products to show more information with images on the product detail page. This guide should give you a 
starting point on how to use media and thumbnails in your Storefront plugin.

## Prerequisites

> In order to build your own subscriber for your plugin, of course you first need a plugin as base. 

Reviewer 4: Der satz passt nicht so recht. wir wollen in diesem artikel ja keinen subscriber bauen, sondern ein storefront plugin...

Author: Jo, ist mein Fehler, falsch übernommen.

To create an own plugin, you can refer to the [Plugin Base Guide](./../plugin-base-guide.md).

Displaying custom images is often done by using custom fields. To take full advantage of this guide, you might want to
read the corresponding guide on [using custom fields](./../administration/add-custom-field.md).

## Using searchMedia function

You're able to store media in your shop and maintaining them in your Administration. It is not possible to display 
such an image in the storefront with only its media ID though. To achieve that, the function `searchMedia` exists:

```php
public function 

> (array $ids, Context $context): MediaCollection { 

Reviewer 4: hier fehlt der funktionsname

... 
}
```

This function reads out the corresponding media objects for the given IDs in order to continue working with them 
afterwards. Here is an example with a custom field (`custom_sports_media_id`) on the product detail page:

> ```html

Author: Maybe because of better phpStorm formatting, I'll straigthen that.

    {% endfor %}

    {# do a single fetch from database #}
    {% set mediaCollection = searchMedia(sportsMediaIds, context.context) %}

    {% for product in searchResult %}
        {# simplify ID access #}
        {% set sportsMediaId = product.translated.customFields.custom_sports_media_id %}

        {# get access to media of product #}
        {% set sportsMedia = mediaCollection.get(sportsMediaId) %}

        {{ dump(sportsMedia) }}
    {% endfor %}
{% endblock %}
```

## Working with sw_thumbnail 

A common issue when developing responsive web pages is resizing images properly for different screen widths. By default,
Shopware generates various thumbnails for each uploaded image. Normally you would have to manually write larger 
chunks of HTML code to render the needed images with `img` and `srcset`.

Fortunately, you do not need to define these attributes on your own - For that, Shopware 
introduced the `sw_thumbnails` Twig function: `sw_thumbnails` automatically generates the `img` and `srcset` code.
This is the minimal configuration:
```twig
{% sw_thumbnails 'my-thumbnails' with {
    media: cover
} %}
```

> As you see, `sw_thumbnail` makes use of one requited parameter: `media` is required and contains the whole media entity.

Reviewer 5: ```suggestion
As you see, `sw_thumbnail` makes use of one required parameter: `media` is required and contains the whole media entity.
```

The string after `sw_thumbnails` is also required but does not render a CSS class. All other parameters are optional.

> Let's have a look on how Shopware uses `sw_thumbnails` inside those

Reviewer 5: inside those... what ?

### Dealing with thumbnail sizes

With the `sizes` parameter you can control the `sizes` Attribute of the `img` and define which of the thumbnails should 
be used in a media query / viewport.

You can find more information on those sizes here:
<!-- markdown-link-check-disable-next-line -->
{% embed url="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-srcset" %}

E.g. if the browser is in Bootstrap viewport `lg` (which is 992px - 1199px) use an image which is closest to 333px.

> If `sizes` is not set shopware will automatically use fallback values from global `shopware.theme.breakpoint`.

Reviewer 5: ```suggestion
If `sizes` is not set, Shopware will automatically use fallback values from global `shopware.theme.breakpoint`.
```

```html
<img 
    src="http://shopware.local/media/06/f0/5c/1614258798/example-image.jpg" 
    srcset="http://shopware.local/media/06/f0/5c/1614258798/example-image.jpg 1921w, 
            http://shopware.local/thumbnail/06/f0/5c/1614258798/example-image_1920x1920.jpg 1920w, 
            http://shopware.local/thumbnail/06/f0/5c/1614258798/example-image_800x800.jpg 800w, 
            http://shopware.local/thumbnail/06/f0/5c/1614258798/example-image_400x400.jpg 400w" 
    sizes="100px">
```

{% hint style="danger" %} Please note that those sizes only work with bootstrap viewports, like xs, sm, md, lg and xl.
Custom media queries will not work. {% endhint %}

### Additional attributes

With the `attributes` param, additional attributes can be applied. Imagine the following example:
```twig
{% sw_thumbnails 'my-thumbnails' with {
    media: cover,
    attributes: {
        'class': 'my-custom-class',
        'alt': 'Doggo with cat',
        'title': 'Doggo with cat'
    }
} %}
```

This will generate the output below:
```html
<img 
    src="..." 
    sizes="..." 
    class="my-custom-class" 
    alt="Doggo with cat" 
    title="Doggo with cat">

> <<<<<<< HEAD

Reviewer 2: what is this?

Author: Looks like Merge Foo

Reviewer 4: rebase foo... needs to be removed
